File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        return daily(heights, bricks, ladders);
    }
private:    
    int daily(vector<int>& heights, int bricks, int ladders) {
        priority_queue<int, vector<int>, greater<int>> pq;
        int n = heights.size();
        
        for(int i = 0; i < n - 1; ++i) {
            int diff = heights[i + 1] - heights[i];
            
            if(diff > 0) 
                pq.push(diff);
            
            if(pq.size() > ladders) {
                bricks -= pq.top();
                pq.pop();
            }
            if(bricks < 0)
                return i;
        }
        return n - 1;
    }
};

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        minHeap = []
        
        n = len(heights)
        
        for i in range(n - 1):
            diff = heights[i + 1] - heights[i]
            
            if diff > 0:
                heappush(minHeap, diff)
            
            if len(minHeap) > ladders:
                bricks -= heappop(minHeap)
            
            if bricks < 0:
                return i
        return n - 1
            